### DISCUSSION TITLE
Role switching setup - user side

### DISCUSSION THREAD

**Discussion by giulianozor** [aws/aws-cli#8683] [Q&A]
Hello,
I might have searched for the wrong terms, but I did fount nothing about that…

scenario:
* one organization where all my users login
* N organizations that my users will access by switching role

Is it possible for an admin to setup the switch role items for the users using cli ? I would like to avoid to make each user to manually go to the switch role page in the dashboard and add all the switch roles for my other organizations…

thanks

**Comment by uldyssian-sh** [comment]
No — there is no supported way for an admin to centrally configure “Switch Role” entries for users via the AWS CLI.

The “Switch Role” entries you see in the AWS Console are a client-side convenience feature. They are stored per user in the browser (local storage / cookies), not as IAM configuration in AWS. Because of that, there is no API or CLI to manage them centrally.

What *can* be centrally managed (and is the correct AWS pattern):

IAM trust and permissions.

You configure this once as an admin:

- In each target account, create an IAM role
- Configure the role’s trust policy to allow the source account (or Identity Center) to assume it
- Attach the appropriate permissions to the role

After that, users are technically allowed to switch roles — the console shortcut is just a UI helper.

How to avoid users manually adding switch roles in the console:

Option 1: Use AWS IAM Identity Center (SSO)  
This is the recommended solution.

- Assign users or groups to accounts and roles centrally
- Users pick the account/role from the SSO portal
- No manual “switch role” setup in the console
- Works across many accounts and organizations

Option 2: Use direct console URLs  
You can generate role-switch URLs and distribute them (docs, bookmarks, internal portal):

https://signin.aws.amazon.com/switchrole?account=<ACCOUNT_ID>&roleName=<ROLE_NAME>&displayName=<LABEL>

This still requires no per-user AWS-side setup.

Option 3: CLI / automation usage  
For CLI and automation, users never need console switch roles at all:

aws sts assume-role \
  --role-arn arn:aws:iam::<ACCOUNT_ID>:role/<ROLE_NAME> \
  --role-session-name my-session

Or via profiles in ~/.aws/config.

Summary:

- “Switch Role” console entries are not centrally configurable (by design)
- There is no CLI or API to manage them
- Central control is achieved through IAM roles + trust policies
- For scale, use IAM Identity Center instead of console switch roles

**Reply by uldyssian-sh** [reply]
Thanks! If this helps, please consider marking it as “Helpful” or “Accepted” 😊
